File: 06resources/solution.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Resources
{
    class MainClass
    {
        public static void Main(string[] args)
        {
            //FakeInput();
            var m = int.Parse(Console.ReadLine());
            var graph = new Dictionary<string, HashSet<string>>();
            for (var i = 0; i < m; i++)
            {
                var edge = Console.ReadLine().Split(' ');
                AddEdge(graph, edge[0], edge[2]);
            }

            var sortedGraph = Sort(graph);
            Console.WriteLine(String.Join(" ", sortedGraph));

        }

        private static List<string> Sort(IDictionary<string, HashSet<string>> graph)
        {
            var used = new HashSet<string>();
            var path = new List<string>();
            while (true)
            {
                var starts = GetStarts(graph, used);
                if (starts.Count == 0)
                {
                    break;
                }

                var start = starts.First();
                path.Add(start);
                used.Add(start);
            }

            return path;
        }

        private static ISet<string> GetStarts(IDictionary<string, HashSet<string>> graph, HashSet<string> used)
        {
            var starts = new SortedSet<string>(
                graph.Keys
                .Where(v => used.Contains(v) == false)
            );

            foreach (var vertex in graph)
            {
                if (used.Contains(vertex.Key))
                {
                    continue;
                }

                foreach (var next in vertex.Value)
                {
                    starts.Remove(next);
                }
            }

            return starts;
        }

        private static void AddEdge(Dictionary<string, HashSet<string>> graph, string v1, string v2)
        {
            if (graph.ContainsKey(v1) == false)
            {
                graph[v1] = new HashSet<string>();
            }

            if (graph.ContainsKey(v2) == false)
            {
                graph[v2] = new HashSet<string>();
            }

            graph[v1].Add(v2);
        }

        public static void FakeInput()
        {
            string input = @"8
index.html needs main.css
main.css needs sub1.css
index.html needs main.js
main.css needs sub2.css
index.html needs logo.png
main.js needs player.png
main.js needs partial.html
partial.html needs partial.js";

            Console.SetIn(new StringReader(input));


        }
    }
}

File: 06resources/solution.js
const getGets = (arr) => {
    let index = 0;

    return () => {
        const toReturn = arr[index];
        index += 1;
        return toReturn;
    };
};
// this is the test
const test =
    `8
index.html needs main.css
main.css needs sub2.css
main.css needs sub1.css
index.html needs main.js
index.html needs logo.png
main.js needs player.png
main.js needs partial.html
partial.html needs partial.js`.split('\n');

const gets = this.gets || getGets(test);
const print = this.print || console.log;
const quit = this.quit || (() => {});

const min = (...values) => {
    let min = values[0];
    for (const value of values) {
        if (value.localeCompare(min) < 0) {
            min = value;
        }
    }

    return min;
};

class Graph {
    constructor() {
        this._nodes = new Map();
    }

    addEdge(from, to) {
        if (!this._nodes.has(from)) {
            this._nodes.set(from, []);
        }

        if (!this._nodes.has(to)) {
            this._nodes.set(to, []);
        }

        this._nodes.get(from).push(to);
    }

    sortNodes() {
        for (const [vertex, nodes] of this._nodes) {
            nodes.sort();
        }
    }

    *[Symbol.iterator]() {
        const dependencies = this._loadDependencies();
        for (const dep of dependencies) {
            yield dep;
        }
    }

    _getStartNodes(used) {
        let startNodes = new Set(this._nodes.keys());
        [...this._nodes].forEach(([vertex, ends]) => {
            if (used.has(vertex)) {
                return;
            }

            ends.forEach((end) => startNodes.delete(end));
        });

        startNodes = [...startNodes].filter((node) => !used.has(node));

        return startNodes;
    }  

    _loadDependencies() {
        const used = new Set();

        const path = [];

        while (true) {
            const startNodes = this._getStartNodes(used);
            if (startNodes.length === 0) {
                break;
            }

            const nextNode = min(...startNodes);
            path.push(nextNode);
            used.add(nextNode);
        }

        return path;
    }
}

const readGraph = () => {
    const graph = new Graph();
    const edgesCount = +gets();
    for (let i = 0; i < edgesCount; i += 1) {
        const [x, y] = gets().split('needs').map((v) => v.trim());
        graph.addEdge(x, y);
    }

    graph.sortNodes();

    return graph;
};

const solve = () => {
    const graph = readGraph();
    return [...graph];
};

print(solve().join(" "));

quit(0);
